presto-spi/src/main/java/com/facebook/presto/spi/relation/LambdaDefinitionExpression.java

```diff
@@ -111,4 +115,55 @@ private static void checkArgument(boolean condition, String message, Object... m
             throw new IllegalArgumentException(format(message, messageArgs));
         }
     }
+    private static class CanonicalizeExpression
+            implements RowExpressionVisitor<String, Void>
+    {
+        private final Map<String, String> canonicalizedArguments = new HashMap<>();
+
+        public CanonicalizeExpression(List<String> arguments, List<Type> argumentTypes)
+        {
+            for (int i = 0; i < arguments.size(); i++) {
+                canonicalizedArguments.put(arguments.get(i), format("%s_%d", argumentTypes.get(i).toString(), i));
```

Comment:
What if someone defines a variable called `bigint_2` in the lambda body. Will that cause a collision?

Reply:
Non of these variable names are going to keep the original form of what the users' specified at this point. So I don't think that would happen. That's said, there's no guarantee it won't happen in the future, so if there are ways to introduce tests to validate these assumptions I'd be happy to add those. We can also add special character like format(%s$%d) so it would not collide.

Reply:
Maybe I am missing some thing, But I thought we cannot (yet) define variable inside lambda body right? 

The only thing I can think is refer other columns inside lambda (a.k.a. lambda capture) . However lambda capture will be desugared during expression optimization (`LambdaCaptureDesugaringRewriter`). We just need to make sure this happens before canonicalization. 

Use special character like `$` is also an standard trick in Presto 😃

Reply:
Yes, this is not defining new variables, it's rewriting reference to input variable to normalize them. For example, if we have two `LambaDefinitionExpression`:
```
lambda1: (expr_1) -> if (expr_1, 1, 2)
lambda2: (expr_2) -> if (expr_2, 1, 2)
```
and both `expr_1` and `expr_2` are of time `BOOLEAN`. The two lambdas are logically the same. They can be compiled to the same bytecode. This change will rewrite both of them to `(boolean_1) -> if (boolean_1, 1, 2)` and the `lambda1.equals(lambda2)` would be `true`.